src/v/datalake/translation/partition_translator_v2.cc

```diff
@@ -0,0 +1,357 @@
+
+    auto translation_f
+      = _translation_ctx
+          ->translate_now(
+            std::move(reader.value()), _inflight_translation_state->as)
+          .finally([this] { return _translation_ctx->flush(); });
+    cancellation_timer.arm(_inflight_translation_state->translate_for);
+    co_await std::move(translation_f).finally([&cancellation_timer] {
+        cancellation_timer.cancel();
+    });
+}
+
+ss::future<> partition_translator::translate_until_stopped() {
+    const auto& id = _data_source->ntp();
+    vassert(
+      _initialized && _scheduler && _reservations,
+      "[{}] Translation started before the translator is properly initialized",
+      id);
+
+    while (!_as.abort_requested()) {
+        retry_chain_node rcn{
+          _as, max_translation_task_timeout, initial_backoff};
+
+        // Reconcile with the coordinator
+        auto result = co_await fetch_latest_translated_offset(rcn);
+        if (result.errc != coordinator::errc::ok) {
+            vlog(_logger.warn, "Failed to fetch translated offset: {}", result);
+            continue;
+        }
+
+        auto last_committed_offset = result.last_iceberg_committed_offset;
+        auto last_translated_offset = result.last_added_offset.value_or(
+          kafka::prev_offset(_data_source->min_offset_for_translation()));
+        auto last_translated_timestamp = result.last_added_timestamp.value_or(
+          model::timestamp::min());
+
+        auto reset_error
+          = co_await _data_source->update_highest_translated_offset(
+            last_translated_offset,
+            last_translated_timestamp,
+            _term,
+            wait_timeout,
+            _as);
+
+        if (reset_error) {
+            vlog(
+              _logger.warn,
+              "error updating highest translated offset: {}, translation will "
+              "be retried",
+              reset_error);
+            continue;
+        }
+
+        // Update partition metrics. Note that last committed offset here
+        // is NOT synchronized with outstanding commit operations. Therefore
+        // if we reach this point before the most recent batch of files has
+        // been committed, the commit lag metric will be out of sync at least
+        // until `wait_for_data_to_translate` wakes up.
+        _data_source->update_commit_lag(last_committed_offset);
+        _data_source->update_translation_lag(last_translated_offset);
+
+        // Wait until some data is ready to be translated.
+        auto begin_offset = co_await _data_source->wait_for_data_to_translate(
+          last_translated_offset, _as);
+
+        // Notify the scheduler that there is some data to translate
+        _scheduler->notify_ready(id);
+
+        auto translate_f = co_await ss::coroutine::as_future<>(
+          translate_when_notified(begin_offset));
+
+        _inflight_translation_state.reset();
+
+        // Let the scheduler know we are done
+        _scheduler->notify_done(id);
```

Comment:
nit: this section would benefit from comment, this seems to be crucial for the interaction between translator and scheduler